src/app/server/Server.cpp

```diff
@@ -191,6 +179,19 @@ CHIP_ERROR Server::Init(const ServerInitParams & initParams)
         SuccessOrExit(err);
     }
 
+    // SetDataModelProvider() actually initializes/starts the provider, which in turn
+    // triggers the initialization of cluster implementations.  We need
+    // to preserve the following ordering guarantees:
+    //
+    // 1) Provider initialization (under SetDataModelProvider) happens after
+    //    SetSafeAttributePersistenceProvider, since the provider can then use
+    //    the safe persistence provider to implement and initialize its own attribute persistence logic.
+    // 2) `InitDataModelHandler` (under SetDataModelProvider), which depends
```

Comment:
InitDataModelHandler is an implementation detail.

The important part is that Start will trigger cluster initialization so _anything a cluster implementation might use_ must be initialized first.  This is the invariant that needs to be documented in the comment.

Putting this before access control init is probably broken.  Putting this before groups init is probably broken.  Putting this before events can be generated is _definitely_ broken.